### PR TITLE
feat(list): Add angular list

### LINKED ISSUE #2245: Add list component
https://pages.github.ibm.com/ai-applications/design/components/list/overview/

### PR DESCRIPTION
**Pull request by Donisius**
Closes #2245 

### Angular Todo:
- [x] Add support for nested list items
- [x] Add support for single select for simple and nested list items
- [x] Add support for multi select for simple and nested list items
- [x] Add secondary value for list item
- [x] Add drag and drop reorder for simple and nested list items
- [x] Add support for row actions
- [x] Add an easy way to pass in context for row actions
- [x] Add documentation
- [x] Make an overridable filter function in the list model
- [x] Add support for different list item sizes
- [x] Add empty state
- [x] Add support for disabled list items
- [x] Find a way to recursively render list items without the need for the `expanded` field in the list model
- [x] Update styles #2601
- [ ] Add more storybook examples
- [ ] Add support for custom list header templates
- [ ] Add support for pagination - Need to add simple pagination
- [ ] Add tests

### Changes in react:

**packages/react/src/components/List/List.jsx**

- Increment nesting level of a list item only if there exists at least one expandable item in the list item's hierarchical level.
    - Based on the specs [here](https://pages.github.ibm.com/ai-applications/design/components/list/overview#categorized-checklist-example-with-pagination) and [here](https://pages.github.ibm.com/ai-applications/design/components/list/overview#hierarchical-list-example-with-search-and-selection)

**packages/react/src/components/List/ListItem/ListItem.jsx**

- Nesting offset now has a width of 32px
    - Based on #2601
- Added disabled classes
    - Based on #2601

### Style changes

Changes to styles are based on #2601

**packages/react/src/components/List/ListItem/_list-item.scss**

- Added focus border for list item
- Added disabled handle
- Added disabled expand
- Added focus border for expand icon
- Updated spacings to match new specs

### CODE REVIEW

**Review by theiliad — APPROVED** [review]

File: packages/angular/src/list/sample-data.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        isDraggable: true,
+      }),
+      new AIListItem({
+        value: 'Vancouver',
+        secondaryValue: '2,581,000',
+        isSelectable: true,
+        isDraggable: true,
+      }),
+    ],
+  }),
```
Comment (cal-smith) [inline packages/angular/src/list/sample-data.ts:47]: Very cool

File: packages/angular/src/list/stories/app-custom-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  includes(searchString: string) {
+    return (
+      (this.secondaryValue !== undefined &&
+        this.secondaryValue !== null &&
+        this.secondaryValue.includes(searchString)) ||
+      this.items.some((listItem) => listItem.includes(searchString))
+    );
+  }
```
Comment (cal-smith) [inline packages/angular/src/list/stories/app-custom-list.component.ts:16]: Neat!

Comment (cal-smith) [reply packages/angular/src/list/stories/app-custom-list.component.ts:16]: Would be cool to have a mixed ListItem demo too

**Review by cal-smith — CHANGES_REQUESTED** [review]
We'll also need a `package.json` in the root of the angular component (check out the other components for what that should look like), so that it builds the list as a separate module.

Otherwise, a few comments, but it's looking good!

Took a quick look at the react changes as well, and they seem reasonable to me, but I definitely want someone more familiar with that to check it over.

File: packages/styles/src/components/GaugeCard/_gauge-card.scss [hunk 3]
```diff
@@ -4,8 +4,10 @@
   display: flex;
   flex-wrap: wrap;
   font-size: carbon--type-scale(1);
-  height: 100%;
-  width: 100%;
+  flex-direction: row;
```
Comment (davidicus) [inline packages/styles/src/components/GaugeCard/_gauge-card.scss:7]: @cal-smith It was my understanding that this package was auto generated from styles in the React package. Are we making edits directly to styles package now? Will the React and Angular styles now be divergent?

Comment (davidicus) [reply packages/styles/src/components/GaugeCard/_gauge-card.scss:7]: or is this the newly generated styles that have been pulled from React package?

**Review by davidicus — APPROVED** [review]
React changes LGTM

File: packages/styles/src/components/GaugeCard/_gauge-card.scss (hunk 3 above)
Comment (cal-smith) [reply packages/styles/src/components/GaugeCard/_gauge-card.scss:7]: I'm guessing it's pulled from styles changed in React, or it hasn't been updated after some changes ... fyi @Donisius

Comment (Donisius) [reply packages/styles/src/components/GaugeCard/_gauge-card.scss:7]: Seems to be pulled over from the changes in react. I ran `sync` and `build` in the styles package, is there anything else that needs to be done (besides `postinstall` in angular) @cal-smith?

Comment (cal-smith) [reply packages/styles/src/components/GaugeCard/_gauge-card.scss:7]: Nope, that's it! This should be good then :+1:

File: packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            (dragOver)="dragOverAbove.emit($event)"
+          ></div>
+          <div
+            aiListTarget
+            targetPosition="below"
+            (dropping)="droppedBelow.emit($event)"
+            (dragOver)="dragOverBelow.emit($event)"
+          ></div>
+        </div>
+        <ng-container [ngTemplateOutlet]="listItem"></ng-container>
```
[the remark this answers is not in the document]
Comment (Donisius) [reply packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts:40]: I don't think having 2 `ng-content`s works (even if they are mutually exclusive like in this case). The last `ng-content` is always the only one that will work.

[the remark this answers is not in the document]
Comment (Donisius) [reply packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts:40]: What we can do is check if wrapper is necessary in `ai-list.component.ts`, if that would be less confusing.

File: packages/angular/src/list/list-item/ai-list-item.class.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
+import { TemplateRef } from '@angular/core';
+
+export class AIListItem {
```
[the remark this answers is not in the document]
Comment (Donisius) [reply packages/angular/src/list/list-item/ai-list-item.class.ts:3]: At the moment all changes do seem to be picked up at least with `expanded`, `selected`, and `indeterminate`. I can add some storybook knobs to test out the other properties and make sure they are also picked up by the change detector.

File: packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cal-smith) [reply packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts:40]: I'm not sure I follow, I'm not suggesting we _add_ an `ng-content` just that we replace the `ng-container` with the `ng-content` from below ... as it stands we're doing exactly that but with more steps

File: packages/angular/src/list/list-item/ai-list-item.class.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cal-smith) [reply packages/angular/src/list/list-item/ai-list-item.class.ts:3]: Sure ... More just something to consider in the event that change detection related bugs crop up

File: packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cal-smith) [reply packages/angular/src/list/list-item/ai-list-item-wrapper.component.ts:40]: Ah I found it, it's hidden above `*ngIf="draggable && !disabled; else listItem"` ... Well that makes more sense.

It does make me wonder if here is the right place to be handling that though, if we only need the wrapper _sometimes_ maybe that switch should be somewhere else? Not something I think we need to fix at the moment, but something to think about ... I'll resolve this!

**Review by cal-smith — APPROVED** [review]

**Pull request merged**